biboumi: init at 6.1

woffs opened the pull request:
TODO: integrate config in NixOS

The config file referenced in the systemd service file is not existent atm and its options should be integrated into configuration.nix, but I am a newbie :-)

###### Motivation for this change

Provide my favourite XMPP IRC gateway.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [x] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

woffs committed:
biboumi: init at 6.1

TODO: integrate config in NixOS

Mic92 commented on pkgs/servers/xmpp/biboumi/default.nix:
```diff
@@ -0,0 +1,29 @@
+    description = "Modern XMPP IRC gateway";
+    platforms = platforms.unix;
+    homepage = https://lab.louiz.org/louiz/biboumi;
+    license = licenses.zlib;
```
Do you want to maintain this?

woffs replied:
I'll add a maintainer line. Would be fine if I'd find a mate here.

Mic92 replied:
Yes and also take a look at `lib/maintainers.nix`

Mic92 commented on pkgs/servers/xmpp/biboumi/default.nix:
```diff
@@ -0,0 +1,29 @@
+  preConfigure = ''
+    grep -lr /etc/biboumi . | while read f
+    do
+      substituteInPlace $f --replace /etc/biboumi $out/etc/biboumi
```
Does makes biboumi reading its configuration always from `$out` or can this be overwritten via command line?

woffs replied:
the config file is a cmdline argument, but "make install" fails without this adjustment.

Mic92 commented on pkgs/servers/xmpp/biboumi/default.nix:
```diff
@@ -0,0 +1,29 @@
+  };
+
+  buildInputs = [ cmake libuuid expat sqlite git libiconv libidn botan2
+    systemd pkgconfig udns ];
```
How is git used during the build?

woffs replied:
Apparently a test suite is fetched. I don't like this, too.

Mic92 replied:
It will also fail to build in sandbox builds, which is what we use on hydra. So this behavior has to be prevented.

woffs replied:
almost done :)

Mic92 replied:
It seems not to download anything during build. So this is fine.

woffs replied:
Fine, but I already had decided to patch it away so I can enable tests. :-)

woffs committed:
biboumi: remove external buildtime dep

- fetch catch.hpp in a reproducible way
- add maintainer
- enable tests
- remove git dep
- enable parallel building
- add pandoc dep for man page
- nitpicks

Mic92 commented:
Only a minor think `/bin/kill` can be replaced here by `${pkgs.procps}/bin/kill`

```console
$ cat lib/systemd/system/biboumi.service
[Unit]
Description=Biboumi, XMPP to IRC gateway
Documentation=man:biboumi(1) https://biboumi.louiz.org
After=network.target
[Service]
Type=notify
ExecStart=/nix/store/8gk42zf0pd2k3p6xbd1yfglv2jqrfl4z-biboumi-6.1/bin/biboumi /nix/store/8gk42zf0pd2k3p6xbd1yfglv2jqrfl4z-biboumi-6.1/etc/biboumi/biboumi.cfg
ExecReload=/bin/kill -s USR1 $MAINPID
WatchdogSec=20
Restart=always
User=nobody
Group=nobody
[Install]
WantedBy=multi-user.target
```

louiz commented:
> This service could be also changed to use the new DynamicUser, but this could actually proposed to upstream.

Would that also work for the case where the process needs to read and write into /var/lib/biboumi, to save its database?
And what are the advantages of DynamicUser?

Mic92 commented:
@louiz no it is not suitable, if a service stores persistent data. I had the intention that biboumi was a pure network service. `DynamicUser` has the advantage over using `nobody`, that no other service/user will have the same uid/gid. Hence you cannot expose data by accident. I would not recommend `nobody` as user, if the service stores persistent data because a different service might be able to read/write to `/var/lib/biboumi`.

Mic92 commented:
For nixos it might be also interesting how many files, the daemon would store in this directory.
We chmod on startup the database directory and allocate static uid/gids for the username, when if a services stores many files (to avoid recursive chmod).

louiz commented:
At the moment, biboumi only store one file here: /var/lib/biboumi/biboumi.sqlite

It’s a file created by sqlite3, storing the whole database.

This is optional (if biboumi is compiled without the sqlite3 support), that’s why it’s nobody:nobody by default, in the case where biboumi doesn’t use any database.

In the future, biboumi might also be able to use postgresql too.

woffs committed:
biboumi: refine substitutions

- only CMakeLists.txt has to be patched regarding /etc/biboumi
- substitute /bin/kill in systemd service file
- prepare for configuring policy_directory in a future cfg file

Mic92 commented:
@louiz I just read that systemd can also chown directories for us with `DynamicUser` http://0pointer.net/blog/dynamic-users-with-systemd.html

louiz commented:
Thank you, that’s very interesting! I’ll probably use that in the upstream biboumi.service file.
